File: Ahorn/entities/bgModeToggle.jl
module BGswitchBGModeToggle

using ..Ahorn, Maple

@mapdef Entity "bgSwitch/bgModeToggle" BGModeToggle(x::Integer, y::Integer, onlyOff::Bool=false, onlyOn::Bool=false, persistent::Bool=false)

const placements = Ahorn.PlacementDict(
    "BG Mode Toggle (On, BGswitch)" => Ahorn.EntityPlacement(
        BGModeToggle,
        "point",
        Dict{String, Any}(
            "onlyOn" => true
        )
    ),
    "BG Mode Toggle (Off, BGswitch)" => Ahorn.EntityPlacement(
        BGModeToggle,
        "point",
        Dict{String, Any}(
            "onlyOff" => true
        )
    ),
    "BG Mode Toggle (Both, BGswitch)" => Ahorn.EntityPlacement(
        BGModeToggle,
    ),
)

function switchSprite(entity::BGModeToggle)
    onlyOff = get(entity.data, "onlyOff", false)
    onlyOn = get(entity.data, "onlyOn", false)
    
    if onlyOff
        return "objects/BGswitch/bgflipswitch/switch13.png"

    elseif onlyOn
        return "objects/BGswitch/bgflipswitch/switch15.png"

    else
        return "objects/BGswitch/bgflipswitch/switch01.png"
    end
end

function Ahorn.selection(entity::BGModeToggle)
    x, y = Ahorn.position(entity)
    sprite = switchSprite(entity)

    return Ahorn.getSpriteRectangle(sprite, x, y)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::BGModeToggle, room::Maple.Room)
    sprite = switchSprite(entity)

    Ahorn.drawSprite(ctx, sprite, 0, 0)
end

end

File: Loenn/entities/bg_mode_toggle.lua
local utils = require("utils")

local bgModeToggle = {}

bgModeToggle.name = "bgSwitch/bgModeToggle"
bgModeToggle.depth = 2000
bgModeToggle.placements = {
    {
        name = "both",
        data = {
            onlyOff = false,
            onlyOn = false,
            persistent = false
        }
    },
    {
        name = "onlyOn",
        data = {
            onlyOff = false,
            onlyOn = true,
            persistent = false
        }
    },
    {
        name = "onlyOff",
        data = {
            onlyOff = true,
            onlyOn = false,
            persistent = false
        }
    }
}

function bgModeToggle.texture(room, entity)
    local onlyOn = entity.onlyOn
    local onlyOff = entity.onlyOff

    if onlyOff then
        return "objects/BGswitch/bgflipswitch/switch13"
    elseif onlyOn then
        return "objects/BGswitch/bgflipswitch/switch15"
    else
        return "objects/BGswitch/bgflipswitch/switch01"
    end
end

function bgModeToggle.selection(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local onlyOff = entity.onlyOff
    
    if onlyOff then
        return utils.rectangle(x - 8, y - 14, 16, 23)
    else
        return utils.rectangle(x - 8, y - 6, 16, 20)
    end
    
end

return bgModeToggle